packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts

```diff
@@ -220,6 +232,14 @@ abstract class PortfolioBase extends cdk.Resource implements IPortfolio {
     AssociationManager.deployWithStackSets(this, product, options);
   }
 
+  public giveAssetBucketsReadToSharedAccounts() {
```

Comment:
We need a test for this new public method.

Reply:
I'd rather not have this method at all. It's just something that can be forgotten to be called. Can we not make sure that if we share products with assets, that the buckets are automatically shared as well?

Reply:
I agree with you, we initially had something along the following:
```
  public shareWithAccount(accountId: string, options: PortfolioShareOptions = {}): void {
    for (const bucket of this.assetBuckets) {
      bucket.grantRead(new iam.AccountPrincipal(accountId));
    };
    const hashId = this.generateUniqueHash(accountId);
    new CfnPortfolioShare(this, `PortfolioShare${hashId}`, {
      portfolioId: this.portfolioId,
      accountId: accountId,
      shareTagOptions: options.shareTagOptions,
      acceptLanguage: options.messageLanguage,
    });
  }
```

 but it's worth posing the question: what if a user creates a portfolio, shares with the accounts first and THEN adds the products to the portfolio. In this case, the asset buckets of the products would not be shared. In order to account for this situation, we can simply wait for the user to finish all their product adds and account shares and then execute the sharing at the very end. If the portfolio had an `onSynth` method or `beforeDeploy`, I agree that the logic would be better suited there (and we could even make this method private), but in order to account for this situation we would have to rely on an Aspect I believe

Reply:
> shares with the accounts first and THEN adds the products to the portfolio. In this case, the asset buckets of the products would not be shared. 

I agree that's a case to watch out for! Well spotted! 

What I don't agree with is the conclusion that then the asset buckets would not be shared. That would seem to be a consequence of how the code is implemented.

What we want is that:

```ts
portfolio.addProduct(X);
portfolio.shareWithAccount(Y);

// and
portfolio.shareWithAccount(Y);
portfolio.addProduct(X);
```

Have the same end result. That seems to be feasible if we do it like this:

```ts
class Portfolio {
  public addProduct(x) {
    for (const account of this.accounts) {
      this.share(x, account);
    }
    this.products.push(x);
  }

  public shareWithAccount(y) {
    for (const product of this.products) {
      this.share(product, y);
    }
    this.accounts.push(y);
  }
}
```

If this code makes you cringe I'm sure you can invent helpers to make it more palatable.

Reply:
Great suggestion! I did cringe a bit, but it appears to be the only way to eliminate the public method and Aspect. Well definitely switch this